packages/flutter_test/lib/src/finders.dart

```diff
@@ -520,121 +907,225 @@ abstract class Finder {
-      return 'exactly one widget with $description$additional: ${widgets.single}';
-    }
-    if (count < 4) {
-      return '$count widgets with $description$additional: $widgets';
-    }
-    return '$count widgets with $description$additional: ${widgets[0]}, ${widgets[1]}, ${widgets[2]}, ...';
+  Iterable<SemanticsNode> get allCandidates {
+    return collectAllSemanticsNodesFrom(root);
   }
 }
 
-/// Applies additional filtering against a [parent] [Finder].
-abstract class ChainedFinder extends Finder {
-  /// Create a Finder chained against the candidates of another [Finder].
-  ChainedFinder(this.parent);
+/// A mixin that applies additional filtering to the results of a parent [Finder].
+ mixin ChainedFinderMixin<CandidateType> on FinderBase<CandidateType> {
 
-  /// Another [Finder] that will run first.
-  final Finder parent;
+  /// Another finder whose results will be further filtered.
+  FinderBase<CandidateType> get parent;
 
   /// Return another [Iterable] when given an [Iterable] of candidates from a
-  /// parent [Finder].
+  /// parent [FinderBase].
   ///
-  /// This is the method to implement when subclassing [ChainedFinder].
-  Iterable<Element> filter(Iterable<Element> parentCandidates);
+  /// This is the method to implement when mixing in [ChainedFinderMixin] or
+  /// extending [ChainedWidgetFinder].
+  Iterable<CandidateType> filter(Iterable<CandidateType> parentCandidates);
 
   @override
-  Iterable<Element> apply(Iterable<Element> candidates) {
-    return filter(parent.apply(candidates));
+  Iterable<CandidateType> findInCandidates(Iterable<CandidateType> candidates) {
+    return filter(parent.findInCandidates(candidates));
   }
 
   @override
-  Iterable<Element> get allCandidates => parent.allCandidates;
+  Iterable<CandidateType> get allCandidates => parent.allCandidates;
 }
 
-class _FirstFinder extends ChainedFinder {
-  _FirstFinder(super.parent);
+/// Applies additional filtering against a [parent] [Finder].
+abstract class ChainedWidgetFinder extends WidgetFinder with ChainedFinderMixin<Element> {
+  /// Create a Finder chained against the candidates of another [Finder].
+  ChainedWidgetFinder(this.parent);
+
+  @override
+  final FinderBase<Element> parent;
+}
 
+mixin _FirstFinderMixin<CandidateType> on ChainedFinderMixin<CandidateType>{
   @override
-  String get description => '${parent.description} (ignoring all but first)';
+  String describeCandidate(Plurality plurality) {
+    return '${parent.describeCandidate(plurality)} (ignoring all but first)';
+  }
 
   @override
-  Iterable<Element> filter(Iterable<Element> parentCandidates) sync* {
+  Iterable<CandidateType> filter(Iterable<CandidateType> parentCandidates) sync* {
     yield parentCandidates.first;
   }
 }
 
-class _LastFinder extends ChainedFinder {
-  _LastFinder(super.parent);
+class _FirstFinder<CandidateType> extends FinderBase<CandidateType>
+  with ChainedFinderMixin<CandidateType>, _FirstFinderMixin<CandidateType> {
```

Comment:
here and elsewhere the indentation is off. I think this needs to be indented by two more to mark it as a continuation of the previous line.

Reply:
Done here and elsewhere.